registry/config.go

```diff
@@ -99,13 +110,38 @@ func newServiceConfig(options ServiceOptions) *serviceConfig {
 		},
 		V2Only: options.V2Only,
 	}
+
+	config.LoadInsecureRegistries(options.InsecureRegistries)
+
+	return config
+}
+
+// LoadInsecureRegistries loads insecure registries to config
+func (config *serviceConfig) LoadInsecureRegistries(registries []string) error {
 	// Split --insecure-registry into CIDR and registry-specific settings.
-	for _, r := range options.InsecureRegistries {
+	for _, r := range registries {
+		// validate insecure registry
+		if _, err := ValidateIndexName(r); err != nil {
+			return err
+		}
 		// Check if CIDR was passed to --insecure-registry
 		_, ipnet, err := net.ParseCIDR(r)
 		if err == nil {
 			// Valid CIDR.
-			config.InsecureRegistryCIDRs = append(config.InsecureRegistryCIDRs, (*registrytypes.NetIPNet)(ipnet))
+			// If ipnet is already in config.InsecureRegistryCIDRs, skip.
+			isFound := false
+			data := (*registrytypes.NetIPNet)(ipnet)
+			for _, value := range config.InsecureRegistryCIDRs {
+				if string(value.IP) == string(data.IP) && string(value.Mask) == string(data.Mask) {
+					isFound = true
+					break
+				}
+			}
+			// ipnet is not found, add it in config.InsecureRegistryCIDRs
+			if !isFound {
+				config.InsecureRegistryCIDRs = append(config.InsecureRegistryCIDRs, data)
```

Comment:
@thaJeztah @aaronlehmann 
Oh, I am afraid race could exist. Since `newRepositoryInfo` would be called via `ResolveRepository`, while `ResolveRepository` would be called via daemon's field `RegistryService`, like `repoInfo, err := daemon.RegistryService.ResolveRepository(ref)`.
Updated it soon.